op_crates/fetch/lib.rs

```diff
@@ -0,0 +1,266 @@
+    // TODO(ry) The Rc below is a hack because we store Rc<CliState> in OpState.
+    // Ideally it could be removed.
+    let permissions = state_.borrow::<Rc<FP>>();
+    permissions.check_net_url(&url_)?;
+  }
+
+  let mut request = client.request(method, url_);
+
+  match data.len() {
+    0 => {}
+    1 => request = request.body(Vec::from(&*data[0])),
+    _ => panic!("Invalid number of arguments"),
+  }
+
+  for (key, value) in args.headers {
+    let name = HeaderName::from_bytes(key.as_bytes()).unwrap();
+    let v = HeaderValue::from_str(&value).unwrap();
+    request = request.header(name, v);
+  }
+  //debug!("Before fetch {}", url);
+
+  let res = request.send().await?;
+
+  //debug!("Fetch response {}", url);
+  let status = res.status();
+  let mut res_headers = Vec::new();
+  for (key, val) in res.headers().iter() {
+    res_headers.push((key.to_string(), val.to_str().unwrap().to_owned()));
+  }
+
+  let rid = state
+    .borrow_mut()
+    .resource_table
+    .add("httpBody", Box::new(res));
+
+  Ok(json!({
+    "bodyRid": rid,
+    "status": status.as_u16(),
+    "statusText": status.canonical_reason().unwrap_or(""),
+    "headers": res_headers
+  }))
+}
+
+pub async fn op_fetch_read(
+  state: Rc<RefCell<OpState>>,
+  args: Value,
+  _data: BufVec,
+) -> Result<Value, AnyError> {
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct Args {
+    rid: u32,
+  }
+
+  let args: Args = serde_json::from_value(args)?;
+  let rid = args.rid;
+
+  use futures::future::poll_fn;
+  use futures::ready;
+  use futures::FutureExt;
+  let f = poll_fn(move |cx| {
+    let mut state = state.borrow_mut();
+    let response = state
+      .resource_table
+      .get_mut::<Response>(rid as u32)
+      .ok_or_else(bad_resource_id)?;
+
+    let mut chunk_fut = response.chunk().boxed_local();
+    let r = ready!(chunk_fut.poll_unpin(cx))?;
+    if let Some(chunk) = r {
+      Ok(json!({ "chunk": &*chunk })).into()
+    } else {
+      Ok(json!({ "chunk": null })).into()
+    }
+  });
+  f.await
+  /*
+  // I'm programming this as I want it to be programmed, even though it might be
+  // incorrect, normally we would use poll_fn here. We need to make this await pattern work.
+  let chunk = response.chunk().await?;
+  if let Some(chunk) = chunk {
+    // TODO(ry) This is terribly inefficient. Make this zero-copy.
+    Ok(json!({ "chunk": &*chunk }))
+  } else {
+    Ok(json!({ "chunk": null }))
+  }
+  */
+}
+
+struct HttpClientResource {
+  client: Client,
+}
+
+impl HttpClientResource {
+  fn new(client: Client) -> Self {
+    Self { client }
+  }
+}
+
+pub fn op_create_http_client<FP>(
+  state: &mut OpState,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<Value, AnyError>
+where
+  FP: FetchPermissions + 'static,
+{
+  #[derive(Deserialize, Default, Debug)]
+  #[serde(rename_all = "camelCase")]
+  #[serde(default)]
+  struct CreateHttpClientOptions {
+    ca_file: Option<String>,
+  }
+
```

Comment:
This is missing a --unstable check.

Reply:
There's no unstable check currently
https://github.com/denoland/deno/blob/cead79f5b8ffd376d339b6e0c30e872bfe6820f6/cli/ops/fetch.rs#L174-L191

Reply:
Oh that is bad - there is meant to be one. `Deno.createHttpClient()` is unstable.